devito/types/basic.py

```diff
@@ -148,99 +143,72 @@ def _C_typedecl(self):
-        return type(name, (cls, ), dict(cls.__dict__))
-
-    def _cached_init(self):
-        """Initialise symbolic object with a cached object state"""
-        original = _SymbolCache[self.__class__]
-        self.__dict__ = original().__dict__
-
-    def __hash__(self):
-        """The hash value of an object that caches on its type is the
-        hash value of the type itself."""
-        return hash(type(self))
-
-
 class AbstractSymbol(sympy.Symbol, Basic, Pickable, Evaluable):
+
     """
-    Base class for dimension-free symbols.
+    Base class for scalar symbols.
 
-    The sub-hierarchy is structured as follows
+    The hierarchy is structured as follows
 
                              AbstractSymbol
                                    |
                  -------------------------------------
                  |                                   |
-        AbstractCachedSymbol                --------------------
-                 |                          |                  |
-              Constant                    Symbol           Dimension
-                                            |
-                                          Scalar
-
-    There are three relevant AbstractSymbol sub-types: ::
-
-        * Symbol: A generic scalar symbol that can be used to build an equation.
-                  It does not carry data. Typically, Symbols are created internally
-                  by Devito (e.g., for temporary variables)
-        * Constant: A generic scalar symbol that can be used to build an equation.
-                    It carries data (a scalar value).
-        * Dimension: A problem dimension, used to create an iteration space. It
-                     may be used to build equations; typically, it is used as
-                     an index for an Indexed.
+             DataSymbol                            Symbol
+                 |                                   |
+         ----------------                   -------------------
+         |              |                   |                 |
+      Constant   DefaultDimension         Scalar          Dimension
+                                                    <:mod:`dimension.py`>
 
-    Notes
-    -----
-    Constants are cached by both SymPy and Devito, while Symbols and Dimensions
-    by SymPy only.
+    All symbols can be used to build equations. However, while DataSymbol
+    carries data, Symbol is a pure symbolic object.
+
+    Constant, DefaultDimension, and Dimension (and most of its subclasses) are
+    part of the user API; Scalar, instead, is only used internally by Devito.
+
+    DefaultDimension and Dimension define a problem dimension (in other words,
+    an "iteration space"). They can be used to index into Functions. For more
+    information, refer to :mod:`dimension.py`.
     """
 
     is_AbstractSymbol = True
+    is_Symbol = True
+
+    @classmethod
+    def _filter_assumptions(cls, **kwargs):
```

Comment:
note for reviewers: some kwargs are for sympy (e.g., assumptions), other for devito (e.g., shape, etc). Here, we split the two